File: day03.hs
import qualified Data.Map.Strict as M
import Data.List (group, sort)
import Data.Bifunctor (bimap)
import Data.Tuple (fst)
import Data.Bits (xor, (.&.))

type Backpack = M.Map Char Int

priority :: Char -> Int
priority = (\x -> x - fromEnum 'A' + 1 - if wasUpper x then 6 else 0) . twiddleCase . fromEnum
  where
    wasUpper :: Int -> Bool
    wasUpper i = i .&. 32 == 32
    
    twiddleCase :: Int -> Int
    twiddleCase = xor 32

clean :: String -> [(Backpack, Backpack)]
clean = map toBackpacks . lines
  where
    toBackpacks :: String -> (Backpack, Backpack)
    toBackpacks = bimap toBackpack toBackpack . halve

    toBackpack :: String -> Backpack
    toBackpack = M.fromAscList . map (\x -> (head x, length x)) . group . sort     

    halve :: String -> (String, String)
    halve s = splitAt half s
      where half = (`div` 2) . length $ s
  
partOne :: [(Backpack, Backpack)] -> Int
partOne = sum . concatMap (map (priority. fst) . M.toList . uncurry M.intersection)

partTwo :: [(Backpack, Backpack)] -> Int
partTwo [] = 0
partTwo (x:y:z:ls) = priority (getBadge [x,y,z]) + partTwo ls 
  where 
    getBadge :: [(Backpack, Backpack)] -> Char
    getBadge = fst . head . M.toList . foldl1 M.intersection . map (uncurry M.union)

main :: IO ()
main = do
  input <- readFile "./input/day03.txt"
  let cleanInput = clean input in do
    print $ partOne cleanInput
    print $ partTwo cleanInput  

File: day03.py
data = ''

with open('./input/day03.txt') as f:
    data = f.read()
    
def chunk(ls, n):
    for i in range(0, len(ls), n):
        yield ls[i:i + n]
    
def split(ls):
    le = len(ls)
    half = int(le / 2)
    return (ls[0: half], ls[half:le])

def first(ls):
    return ls[0]

def score(ch):
    ich = ord(ch)
    return ich - 65 + 27 if ich & 32 == 0 else ich - 97 + 1 

def score_list(ls):
    return sum(map(lambda x: score(first(x)), ls))

processed = [line for line in data.split("\n") if line.strip() != '']
splitted = map(split, processed)
chunked = chunk(processed, 3);

part_one = score_list([[z for z in x if z in y] for (x, y) in splitted]) 
part_two = score_list([[ch for ch in x if ch in y and ch in z] for [x, y, z] in chunked])

print(part_one)
print(part_two)